File: javascript/src/llm.js
const MAX_RESPONSE_BYTES = 1_048_576;
const SAFE_ERROR = Symbol('sanitized LLM error');

function failure(code, message, status) {
  return Object.assign(new Error(message), { code, [SAFE_ERROR]: true, ...(status === undefined ? {} : { status }) });
}

function integer(value, name, min, max) {
  if (!Number.isInteger(value) || value < min || value > max) {
    throw failure('LLM_CONFIG', `${name} must be an integer between ${min} and ${max}.`);
  }
}

function endpoint(baseUrl) {
  let url;
  try { url = new URL(baseUrl); } catch { throw failure('LLM_CONFIG', 'Invalid LLM base URL.'); }
  const loopback = ['localhost', '127.0.0.1', '[::1]'].includes(url.hostname);
  if ((url.protocol !== 'https:' && !(url.protocol === 'http:' && loopback)) || url.username || url.password || url.search || url.hash) {
    throw failure('LLM_CONFIG', 'LLM URL must use HTTPS (HTTP only for localhost), without credentials, query, or fragment.');
  }
  url.pathname = `${url.pathname.replace(/\/+$/, '')}/chat/completions`;
  return url.toString();
}

function aborted() { return failure('LLM_ABORTED', 'LLM request was cancelled.'); }

async function delay(milliseconds, signal) {
  if (signal?.aborted) throw aborted();
  await new Promise((resolve, reject) => {
    const done = () => { signal?.removeEventListener('abort', cancel); resolve(); };
    const timer = setTimeout(done, milliseconds);
    const cancel = () => { clearTimeout(timer); signal?.removeEventListener('abort', cancel); reject(aborted()); };
    signal?.addEventListener('abort', cancel, { once: true });
  });
}

function retryDelay(header, attempt) {
  if (header) {
    const seconds = Number(header);
    const milliseconds = Number.isFinite(seconds) ? seconds * 1000 : Date.parse(header) - Date.now();
    if (Number.isFinite(milliseconds)) return Math.max(0, Math.min(milliseconds, 5000));
  }
  return Math.min(250 * 2 ** attempt, 2000);
}

async function readJSON(response) {
  const declaredSize = Number(response.headers.get('content-length'));
  if (declaredSize > MAX_RESPONSE_BYTES) {
    await response.body?.cancel().catch(() => {});
    throw failure('LLM_PROTOCOL', 'LLM response exceeded the size limit.');
  }
  if (!response.body) throw failure('LLM_PROTOCOL', 'LLM response body was empty.');
  const reader = response.body.getReader();
  const chunks = [];
  let length = 0;
  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      length += value.byteLength;
      if (length > MAX_RESPONSE_BYTES) {
        await reader.cancel().catch(() => {});
        throw failure('LLM_PROTOCOL', 'LLM response exceeded the size limit.');
      }
      chunks.push(value);
    }
  } finally { reader.releaseLock(); }
  const bytes = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) { bytes.set(chunk, offset); offset += chunk.byteLength; }
  try { return JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(bytes)); }
  catch { throw failure('LLM_PROTOCOL', 'LLM returned invalid JSON.'); }
}

/** A small real HTTP client; it contains no planning, tools, memory, or Agent framework. */
export class ChatClient {
  constructor({ apiKey, provider = 'qwen', baseUrl = provider === 'qwen' ? 'https://dashscope.aliyuncs.com/compatible-mode/v1' : 'https://api.openai.com/v1', model = provider === 'qwen' ? 'qwen-plus' : 'gpt-4.1-mini', timeoutMs = 30000, maxRetries = 2, fetchImpl = globalThis.fetch, maxOutputTokens = 1200 } = {}) {
    if (typeof apiKey !== 'string' || !apiKey.trim() || /[\r\n]/.test(apiKey)) throw failure('LLM_CONFIG', 'A nonempty server-side LLM API key is required.');
    if (typeof model !== 'string' || !model.trim() || model.length > 200) throw failure('LLM_CONFIG', 'A valid LLM model is required.');
    if (!['qwen', 'openai'].includes(provider)) throw failure('LLM_CONFIG', 'LLM provider must be qwen or openai.');
    integer(timeoutMs, 'timeoutMs', 1, 120000);
    integer(maxRetries, 'maxRetries', 0, 5);
    integer(maxOutputTokens, 'maxOutputTokens', 1, 128000);
    if (typeof fetchImpl !== 'function') throw failure('LLM_CONFIG', 'fetchImpl must be a function.');
    // Keep credentials non-enumerable so accidental JSON serialization cannot disclose them.
    Object.defineProperty(this, 'apiKey', { value: apiKey.trim(), enumerable: false });
    Object.assign(this, { endpoint: endpoint(baseUrl), model, provider, timeoutMs, maxRetries, fetchImpl, maxOutputTokens });
  }

  async complete({ messages, tools = [], signal } = {}) {
    if (!Array.isArray(messages) || !messages.length || !Array.isArray(tools)) throw failure('LLM_CONFIG', 'messages must be nonempty and tools must be an array.');
    const body = { model: this.model, messages, stream: false };
    if (this.provider === 'qwen') {
      body.max_tokens = this.maxOutputTokens;
      body.enable_thinking = false;
    } else { body.max_completion_tokens = this.maxOutputTokens; }
    if (tools.length) {
      body.tools = tools.map(tool => {
        if (this.provider !== 'qwen' || !tool?.function) return tool;
        const { strict: _strict, ...functionSchema } = tool.function;
        return { ...tool, function: functionSchema };
      });
      body.tool_choice = 'auto';
      body.parallel_tool_calls = false;
    }
    let encoded;
    try { encoded = JSON.stringify(body); } catch { throw failure('LLM_CONFIG', 'LLM request could not be serialized.'); }
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (signal?.aborted) throw aborted();
      const controller = new AbortController();
      let timedOut = false;
      const timeout = setTimeout(() => { timedOut = true; controller.abort(); }, this.timeoutMs);
      const cancel = () => controller.abort();
      signal?.addEventListener('abort', cancel, { once: true });
      let retryAfter;
      let currentError;
      let retryable = false;
      try {
        const response = await this.fetchImpl(this.endpoint, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${this.apiKey}` },
          body: encoded,
          signal: controller.signal,
          redirect: 'error'
        });
        if (!response.ok) {
          retryAfter = response.headers.get('retry-after');
          await response.body?.cancel().catch(() => {});
          retryable = response.status === 429 || response.status >= 500;
          const code = response.status === 401 || response.status === 403 ? 'LLM_AUTH' : response.status === 429 ? 'LLM_RATE_LIMIT' : 'LLM_HTTP';
          throw failure(code, `LLM request failed with HTTP ${response.status}.`, response.status);
        }
        return await readJSON(response);
      } catch (err) {
        if (signal?.aborted) throw aborted();
        if (timedOut) {
          currentError = failure('LLM_TIMEOUT', 'LLM request timed out.');
          retryable = true;
        } else if (err?.[SAFE_ERROR] === true) {
          currentError = err;
        } else {
          // Never copy a network error message or provider body: either could contain secrets.
          currentError = failure('LLM_NETWORK', 'LLM network request failed.');
          retryable = true;
        }
      } finally {
        clearTimeout(timeout);
        signal?.removeEventListener('abort', cancel);
      }
      if (!retryable || attempt === this.maxRetries) throw currentError;
      await delay(retryDelay(retryAfter, attempt), signal);
    }
  }
}

File: python/agent/llm.py
"""Real standard-library HTTP client with bounded retries and sanitized errors.

transport(url, *, headers, body, timeout) may be sync or async and returns
(status, response_headers, body_bytes). timeout and injected sleep use seconds.
The default urllib transport runs in a worker thread. asyncio timeout/cancellation
discards its late result; Python cannot forcibly stop that worker thread or revoke
an HTTP request already accepted by the provider.
"""
import asyncio
from collections.abc import Mapping
from copy import deepcopy
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
import inspect
import json
import math
import socket
import time
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit, urlunsplit
from urllib.request import HTTPRedirectHandler, Request, build_opener

from .config import Config
from .errors import AgentError
from .jsonutil import dumps

MAX_RESPONSE_BYTES = 1_048_576


class _ResponseTooLarge(Exception):
    pass


class _NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def _endpoint(base_url):
    try:
        if not isinstance(base_url, str) or any(ord(char) < 32 for char in base_url):
            raise ValueError
        url = urlsplit(base_url)
        hostname = url.hostname
        _ = url.port  # Validate a malformed/non-numeric port before issuing HTTP.
        loopback = hostname in ("localhost", "127.0.0.1", "::1")
        if not hostname or (url.scheme != "https" and not (url.scheme == "http" and loopback)) or url.username is not None or url.password is not None or url.query or url.fragment:
            raise ValueError
        return urlunsplit((url.scheme, url.netloc, url.path.rstrip("/") + "/chat/completions", "", ""))
    except (ValueError, TypeError):
        raise AgentError("LLM_CONFIG", "LLM URL must use HTTPS (HTTP only for localhost), without credentials, query, or fragment.") from None


def _integer(value, name, minimum, maximum):
    if type(value) is not int or not minimum <= value <= maximum:
        raise AgentError("LLM_CONFIG", f"{name} must be an integer between {minimum} and {maximum}.")


def _default_transport(url, *, headers, body, timeout):
    deadline = time.monotonic() + timeout
    request = Request(url, data=body, headers=headers, method="POST")
    try:
        response = build_opener(_NoRedirect()).open(request, timeout=timeout)
    except HTTPError as error:
        # Error bodies are neither read nor propagated, including redirect responses.
        with error:
            return error.code, dict(error.headers.items()), b""
    with response:
        response_headers = dict(response.headers.items())
        try:
            declared_size = int(response.headers.get("Content-Length", "0"))
        except ValueError:
            declared_size = 0
        if declared_size > MAX_RESPONSE_BYTES:
            raise _ResponseTooLarge()
        chunks, size = [], 0
        read = getattr(response, "read1", response.read)
        while True:
            if time.monotonic() >= deadline:
                raise TimeoutError()
            chunk = read(min(65536, MAX_RESPONSE_BYTES + 1 - size))
            if not chunk:
                break
            chunks.append(chunk)
            size += len(chunk)
            if size > MAX_RESPONSE_BYTES:
                raise _ResponseTooLarge()
        return response.status, response_headers, b"".join(chunks)


def _retry_delay(header, attempt):
    if header:
        try:
            seconds = float(header)
            if math.isfinite(seconds):
                return min(5.0, max(0.0, seconds))
        except (ValueError, TypeError):
            pass
        try:
            when = parsedate_to_datetime(header)
            if when.tzinfo is None:
                when = when.replace(tzinfo=timezone.utc)
            return min(5.0, max(0.0, (when - datetime.now(timezone.utc)).total_seconds()))
        except (ValueError, TypeError, OverflowError):
            pass
    return min(0.25 * 2 ** attempt, 2.0)


def _reject_constant(_value):
    raise ValueError()


def _finite_float(value):
    number = float(value)
    if not math.isfinite(number):
        raise ValueError()
    return number


class ChatClient:
    def __init__(self, config: Config, transport=None, sleep=None):
        if not isinstance(config.api_key, str) or not config.api_key.strip() or "\r" in config.api_key or "\n" in config.api_key:
            raise AgentError("LLM_CONFIG", "A nonempty server-side LLM API key is required.")
        if config.provider not in ("qwen", "openai"):
            raise AgentError("LLM_CONFIG", "LLM provider must be qwen or openai.")
        if not isinstance(config.model, str) or not config.model.strip() or len(config.model) > 200:
            raise AgentError("LLM_CONFIG", "A valid LLM model is required.")
        _integer(config.timeout_ms, "timeout_ms", 1, 120000)
        _integer(config.max_retries, "max_retries", 0, 5)
        _integer(config.max_output_tokens, "max_output_tokens", 1, 128000)
        if transport is not None and not callable(transport):
            raise AgentError("LLM_CONFIG", "transport must be callable.")
        if sleep is not None and not callable(sleep):
            raise AgentError("LLM_CONFIG", "sleep must be callable.")
        self.endpoint = _endpoint(config.base_url)
        self.provider, self.model = config.provider, config.model
        self.timeout_ms, self.max_retries = config.timeout_ms, config.max_retries
        self.max_output_tokens = config.max_output_tokens
        self._api_key = config.api_key.strip()
        self._transport = transport or _default_transport
        self._sleep = sleep or asyncio.sleep

    async def _request(self, body):
        kwargs = {
            "headers": {"Content-Type": "application/json", "Accept-Encoding": "identity", "Authorization": f"Bearer {self._api_key}"},
            "body": body, "timeout": self.timeout_ms / 1000,
        }
        if inspect.iscoroutinefunction(self._transport):
            return await self._transport(self.endpoint, **kwargs)
        result = await asyncio.to_thread(self._transport, self.endpoint, **kwargs)
        return await result if inspect.isawaitable(result) else result

    async def complete(self, *, messages, tools=None):
        if tools is None:
            tools = []
        if not isinstance(messages, list) or not messages or not isinstance(tools, list):
            raise AgentError("LLM_CONFIG", "messages must be nonempty and tools must be an array.")
        body = {"model": self.model, "messages": messages, "stream": False}
        if self.provider == "qwen":
            body.update(max_tokens=self.max_output_tokens, enable_thinking=False)
        else:
            body["max_completion_tokens"] = self.max_output_tokens
        try:
            if tools:
                body["tools"] = deepcopy(tools)
                if self.provider == "qwen":
                    for tool in body["tools"]:
                        if isinstance(tool, dict) and isinstance(tool.get("function"), dict):
                            tool["function"].pop("strict", None)
                body.update(tool_choice="auto", parallel_tool_calls=False)
            encoded = dumps(body).encode("utf-8")
        except (TypeError, ValueError, RecursionError, UnicodeError):
            raise AgentError("LLM_CONFIG", "LLM request could not be serialized.") from None
        for attempt in range(self.max_retries + 1):
            retryable, retry_after = False, None
            try:
                raw = await asyncio.wait_for(self._request(encoded), self.timeout_ms / 1000)
            except asyncio.CancelledError:
                raise
            except (TimeoutError, socket.timeout):
                error = AgentError("LLM_TIMEOUT", "LLM request timed out.")
                retryable = True
            except _ResponseTooLarge:
                error = AgentError("LLM_PROTOCOL", "LLM response exceeded the size limit.")
            except Exception as caught:
                timed_out = isinstance(caught, URLError) and isinstance(caught.reason, (TimeoutError, socket.timeout))
                error = AgentError("LLM_TIMEOUT" if timed_out else "LLM_NETWORK", "LLM request timed out." if timed_out else "LLM network request failed.")
                retryable = True
            else:
                if not isinstance(raw, (tuple, list)) or len(raw) != 3:
                    raise AgentError("LLM_PROTOCOL", "LLM transport returned an invalid response.")
                status, headers, response_body = raw
                if type(status) is not int or not 100 <= status <= 599 or not isinstance(headers, Mapping):
                    raise AgentError("LLM_PROTOCOL", "LLM transport returned an invalid response.")
                headers = {str(name).lower(): str(value) for name, value in headers.items()}
                if not 200 <= status < 300:
                    retry_after = headers.get("retry-after")
                    retryable = status == 429 or status >= 500
                    code = "LLM_AUTH" if status in (401, 403) else "LLM_RATE_LIMIT" if status == 429 else "LLM_HTTP"
                    error = AgentError(code, f"LLM request failed with HTTP {status}.", status=status)
                else:
                    try:
                        declared_size = int(headers.get("content-length", "0"))
                    except ValueError:
                        declared_size = 0
                    if not isinstance(response_body, bytes) or len(response_body) > MAX_RESPONSE_BYTES or declared_size > MAX_RESPONSE_BYTES:
                        raise AgentError("LLM_PROTOCOL", "LLM response is invalid or exceeded the size limit.")
                    try:
                        parsed = json.loads(response_body.decode("utf-8"), parse_constant=_reject_constant, parse_float=_finite_float)
                    except (UnicodeError, ValueError, RecursionError):
                        raise AgentError("LLM_PROTOCOL", "LLM returned invalid JSON.") from None
                    if not isinstance(parsed, dict):
                        raise AgentError("LLM_PROTOCOL", "LLM response must be a JSON object.")
                    return parsed
            if not retryable or attempt == self.max_retries:
                raise error from None
            delayed = self._sleep(_retry_delay(retry_after, attempt))
            if inspect.isawaitable(delayed):
                await delayed
